Shopify/cli
# allow dev dash apps to have optional modules

### WHY are these changes introduced?

Fixes: https://github.com/Shopify/develop-app-inner-loop/issues/2290
- Our App schema currently requires application_url, embedded and the webhooks api_version to always be set
- This is not the case with app management apps, and will eventually never be the case with the greenfield vision of all app modules being optional

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/183IglXPcqlyh0IdGCI5/f4584ebb-c05a-4146-b4a9-bc2e7bb3a039.png)

### WHAT is this pull request doing?

- introduce a new specification schema modifier method `customizeSchemaForDevPlatformClient` that is run when fetching the App schema
- this merges any overrides into the schema, and is needed as another specification schema modifier method `contributeToAppConfigurationSchema` reduces its own changes on the base AppSchema, so we need to append onto this after as well

- I discussed with @shauns and making the partners client the "special" case seemed like the right idea, as eventually we'll be ripping all this out when all modules/specs are all optional.

- Due to the partner client dependency, I've had to "drill" the client name down quite a few levels, and default to "partners" on some occasions we don't have access to the org yet. I think this is fine, and I've been told to take the "drill" approach before, but please lmk what we think

### How to test your changes?

- you can link a management app that is missing `application_url`, `embedded` or `webhooks`
- you can create and deploy a management app missing these fields
- with a partners app, everything should operate as before, where all of these fields are required


[Dev Dash app spin link](https://dev.shopify.shopify-developer-dashboard-01bm.alex-montague.us.spin.dev/dashboard/2/apps/905743398573/versions)
[Partners app spin link](https://partners.shopify-developer-dashboard-01bm.alex-montague.us.spin.dev/1/apps/905743398574/edit)


#### 🎩  Partners App:

```toml
# partners app toml
# any missing application_url, embedded or webhooks throws error still
# Learn more about configuring your app at https://shopify.dev/docs/apps/tools/cli/configuration

client_id = "5d44973d3cae826954b02396337f60fa"
name = "partners-spin-cli-app"
handle = "partners-spin-cli-app"

[webhooks]
api_version = "unstable"

[access_scopes]
# Learn more at https://shopify.dev/docs/apps/tools/cli/configuration#access_scopes
scopes = ""

[auth]
redirect_urls = [ "https://shopify.dev/apps/default-app-home/api/auth" ]

[pos]
embedded = false
```

on link:

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/183IglXPcqlyh0IdGCI5/35e9334a-f375-4f6f-bf56-c31900f7ac6c.png)

on deploy:

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/183IglXPcqlyh0IdGCI5/ebdc9357-ee2e-4f7c-a513-4fa4bd5694c2.png)


#### 🎩  Management App:

```toml
# management app toml
# missing application_url, embedded, or webhooks fields are fine
# Learn more about configuring your app at https://shopify.dev/docs/apps/tools/cli/configuration

client_id = "008494d258802e2f1540d502316b4637"
app_id = "gid://shopify/App/905743398573"
organization_id = "2"
name = "my empty spin app"

[access_scopes]
# Learn more at https://shopify.dev/docs/apps/tools/cli/configuration#access_scopes
scopes = ""

[auth]
redirect_urls = [ ]
```

on link:

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/183IglXPcqlyh0IdGCI5/054f30c9-2cf4-4e93-9e15-364dedd31480.png)

on deploy:

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/183IglXPcqlyh0IdGCI5/5343bd36-309b-4ba6-b4ba-75527600fe48.png)
